src/Stratis.Bitcoin.Features.MemoryPool/MempoolBehavior.cs

```diff
@@ -508,14 +525,30 @@ public Task RelayTransaction(uint256 hash)
             // find all behaviours then start an exclusive task
             // to add the hash to each local collection
             IEnumerable<MempoolBehavior> behaviours = peers.Select(s => s.Behavior<MempoolBehavior>());
-            this.logger.LogTrace("(-)");
-            return this.manager.MempoolLock.WriteAsync(() =>
+            return this.manager.MempoolLock.WriteAsync(() => 
```

Comment:
Ah but we need a global lock hmmm...

I suggest if we don't have a global lock anymore and we use `Hashset<>` then perhaps create a new method on each behaviour that will do the following code inside the lock lock.
```
 if (!mempoolBehavior.filterInventoryKnown.ContainsKey(hash))
    mempoolBehavior.inventoryTxToSend.TryAdd(hash, hash);
```

Also try to keep tracing outside of the locks.

Reply:
If I remove the mempool manager locking here I had issues with all the peers getting dropped with "pong timeouts".

Reply:
Hu can you elaborate? 
If we need a video chat let me know.

Reply:
If you removed the global lock then you will get race conditions as each behaviour will try to do the same (it might be you got internal exceptions that where not logged) 🤔 

You have to either use a global lock (global to all behaviours) or lock each behaviour separately (move the code to a method and put a lock around it and call that method for each behaviour).   

if you use CD then you wont have an issue, but @Aprogiena might argue against CD even though I think this is quite a simple use of collections to justify it (unless you use hashset).

Reply:
Ok that makes sense. I'll wait to see Apro's opinion and if he doesn't recommend something else go with the CD and see how that works. Thanks!

Reply:
Leave it for now as you have it, it will be fixed in the mempool behavior refactoring later on. 

As mentioned before- use lock keyworkd and hashset for the inner datastructures.

Reply:
What do you mean leave it for now? using the mempool lock?
I prefer if we did not leave it and try to fix it in this PR.

I suggest to create new method on the behavior that will lock the hashset

Reply:
The suggestion is to leave that one use of the original lock and attached as Mike reported that there is some problem with that. There are 20 other problems with this code and it is out of scope of this PR to fix them. You would basically need to completely rewrite the whole thing. I believe we can take this small PR which improves the situation slightly but also leaves many problems unresolved.

Reply:
Indeed there are but I want to get rid of the scheduler lock and I believe this can be done, so @mikedennis can you please attempt to fix this as I described earlier? (each behaviour will do its own locking on its dictionaries) and see if you still get nodes dropping.

Reply:
I will give it another try and see if I can get rid of this one as well with the strategy you suggested. Thanks!